core/src/Cuda/Kokkos_Cuda_Vectorization.hpp

```diff
@@ -70,6 +70,22 @@ struct in_place_shfl_op {
     return *static_cast<Derived const*>(this);
   }
 
+  // sizeof(Scalar) < sizeof(int) case
+  template <class Scalar>
+  // requires _assignable_from_bits<Scalar>
+  __device__ inline typename std::enable_if<sizeof(Scalar) < sizeof(int)>::type
+  operator()(Scalar& out, Scalar const& in, int lane_or_delta, int width,
+             unsigned mask = shfl_all_mask) const noexcept {
+    using conv_type = int;
+    conv_type tmp   = static_cast<conv_type>(in);
```

Comment:
Just saying this again. Don't we need to do something like a union here or so? The static_cast uses conversion mechanisms doesn't it? So if we for example convert a half precision float to int we are in trouble:
```c++
#include<cstdio>
int main() {
  float a;
  long long b;
  a = 1.3;
  b = static_cast<long long>(a);
  float c = static_cast<float>(b);
  printf("%f %li %f\n",a,b,c);
}
```
```
g++ main.cpp
./a.out
1.300000 1 1.000000
```

A union should work:
```c++
#include<cstdio>

union conv_type{
  long long val;
  float org_val;
};

int main() {
  float a;
  long long b;
  a = 1.3;
  b = static_cast<long long>(a);
  float c = static_cast<float>(b);
  printf("%f %li %f\n",a,b,c);

  conv_type conv;
  conv.org_val = a;
  conv_type conv2;
  conv2.val = conv.val;
  printf("%f %li %f\n",conv.org_val,conv.val,conv2.org_val);

}
```
```
g++ main.cpp
./a.out
1.300000 1 1.000000
1.300000 1067869798 1.300000
```

Reply:
This approach has been integrated into the template specialization, and the code for size<sizeof(int) and size==sizeof(int) were merged together.  

I also modified the test template to take the data type, then I added a float and double test.  

The float test exercises the new template logic, and the double test found an issue with the mask constant.